### PR TITLE
[1pt] PR: Manually calibrate synthetic rating curves

### LINKED ISSUE #944: [21pt] Manual calibration of Pittsburgh reaches
The upper reaches of HUC `05030101` near Pittsburgh are over-inundating lower flows. This is a major concern for field personnel and the upcoming public FIM release. We need to come up with a short-term, local solution until a more permanent and widespread fix can be built (e.g. Arash's transect ML routine).

@RyanSpies-NOAA and I have brainstormed and have concluded that the best course of action would be to "calibrate" the affected feature_ids by manually providing calibration coefficients to decrease Manning's n thereby narrowing flood stage and extent at lower flows. Another possible option would be for us to obtain cross-section or bathymetry data and use it to estimate cross-sectional area lost to bathymetry (I have reached out to the field to see if we can find some of this data).

You will probably need to insert this solution into the post-processing routine where calibration occurs. Ideally, the solution should be reproducible for other problem areas. I was thinking something like a CSV file that can be read in during post-processing that has columns: HUC8, feature_id, and calibration_coefficient. Feel free to explore other possible solutions and contact Ryan if you have any questions regarding calibration.

### PR DESCRIPTION
**Pull request by mluck**
Adds a feature to manually calibrate rating curves for specified NWM `feature_id`s using a CSV of manual coefficients to output a new rating curve. Manual calibration is applied after any/all other calibrations. Coefficient values between 0 and 1 increase the discharge value (and decrease inundation) for each stage in the rating curve while values greater than 1 decrease the discharge value (and increase inundation).

Manual calibration is performed if `manual_calb_toggle="True"` and the file specified by `man_calb_file` (with `HUC8`, `feature_id`, and `calb_coef_manual` fields) exists. The original HUC-level `hydrotable.csv` (after calibration) is saved with a suffix of `_pre-manual` before the new rating curve is written. Closes #944.

### Additions

- `src/src_manual_calibration.py`: Adds functionality for manual calibration by CSV file

### Changes

- `config/params_template.env`: Adds `manual_calb_toggle` and `man_calb_file` parameters
- `fim_post_processing.sh`: Adds check for toggle and if `man_calb_file` exists before running manual calibration

### Testing
Ran `fim_pipeline.sh`, `tools/inundate_gms.Inundate_gms()`, and `tools/mosaic_inundation.Mosaic_inundation()` on HUCs to produce output example in https://github.com/NOAA-OWP/inundation-mapping/issues/944#issuecomment-1650148029.

### Note

This feature branch depends on the bug fixes in #950 so `dev-fix-aggregated-hydrotable-inundation` is merged into this code.

### Issuer Checklist (For developer use)

_You may update this checklist before and/or after creating the PR. If you're unsure about any of them, please ask, we're here to help! These items are what we are going to look for before merging your code._

- [x] Informative and human-readable title, using the format: `[_pt] PR: <description>`
- [x] Links are provided if this PR resolves an issue, or depends on another other PR
- [x] If submitting a PR to the `dev` branch (the default branch), you have a descriptive [Feature Branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) name using the format: `dev-<description-of-change>` (e.g. `dev-revise-levee-masking`)
- [x] Changes are limited to a single goal (no scope creep)
- [x] The feature branch you're submitting as a PR is up to date (merged) with the latest `dev` branch
- [x] Changes adhere to [PEP-8 Style Guidelines](https://peps.python.org/pep-0008/)
- [x] Any _change_ in functionality is tested
- [x] Passes all unit tests locally (inside interactive Docker container, at `/foss_fim/`, run: `pytest unit_tests/`)
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [x] [CHANGELOG](/docs/CHANGELOG.md) updated with template version number, e.g. `4.x.x.x` 
- [x] [Reviewers requested](https://help.github.com/articles/requesting-a-pull-request-review/)
- [x] Add yourself as an [assignee](https://docs.github.com/en/issues/tracking-your-work-with-issues/assigning-issues-and-pull-requests-to-other-github-users) in the PR  as well as the FIM Technical Lead

### Merge Checklist (For Technical Lead use only)

- [ ] Update [CHANGELOG](/docs/CHANGELOG.md) with latest version number
- [ ] Update the [Citation.cff](/CITATION.cff) file to reflect the latest version number in the [CHANGELOG](/docs/CHANGELOG.md)
- [ ] If applicable, update [README](/README.md) with major alterations

### CODE REVIEW

**Review by robgpita-noaa — CHANGES_REQUESTED** [review]
Nice work on this one @mluck!!

After running `fim_pipeline.sh -n robgp_test_pr_952 -u "05030101 05020005 05010009" -o -jh 1 -jb 6`:
 
I was able to successfully run `src_manual_calibration.py`:
```
python3 /foss_fim/src/src_manual_calibration.py -fim_dir "/data/outputs/robgp_test_pr_952"
    -calb_file "/data/inputs/rating_curve/manual_calibration_coefficients.csv"
```

The contents of `manual_calibration_coefficients.csv` were as follows:

```
HUC8,feature_id,calb_coef_manual
05030101,3824135,0.50
05030101,3824131,0.50
05030101,11050844,0.50
05030101,11050846,0.50
05030101,3786893,0.50
05030101,3786927,0.50
05020005,3786927,0.50
05010009,11050844,0.50
05010009,11050846,0.50
```

I did not inspect the updated hydrotables, but can confirm certain lines (I can only assume those lines are the branches provided in `manual_calibration_coefficients.csv`) of the `hydrotable.csv` have been updated. The `grep` command shows a difference of 731 lines, and the `wc -l` command shows the total lines in each file.

```
$ grep -wvFf /data/outputs/robgp_test_pr_952/05030101/hydrotable_pre-manual.csv /data/outputs/robgp_test_pr_952/05030101/hydrotable.csv | wc -l
731
$ wc -l /data/outputs/robgp_test_pr_952/05030101/hydrotable.csv
374893 /data/outputs/robgp_test_pr_952/05030101/hydrotable.csv
$wc -l /data/outputs/robgp_test_pr_952/05030101/hydrotable_pre-manual.csv
374473 /data/outputs/robgp_test_pr_952/05030101/hydrotable_pre-manual.csv
``````

File: src/src_manual_calibration.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+        else:
+            raise ValueError(f'Manual calibration coefficients must be greater than 0. Minimum value found: {manual_calib_df["calb_coef_manual"].min()}')
+
+    else:
+        raise FileNotFoundError(f'No calibration file found at {calibration_file}. Skipping manual calibration.')
+
+
+if __name__ == '__main__':
+    ## Parse arguments.
```
[the remark this answers is not in the document]
Comment (mluck) [reply src/src_manual_calibration.py:130]: I added a usage example as a comment on the bottom of `src/src_manual_calibration.py`. That script is automatically run in https://github.com/NOAA-OWP/inundation-mapping/blob/dev-manual-calibration/fim_post_processing.sh#L155 if `manual_calb_toggle="True"` is set in `config/params_template.env` and the CSV with the manual calibration coefficients exists.

**Review by robgpita-noaa — APPROVED** [review]

**Comment by robgpita-noaa** [comment]
@mluck Thanks for addressing my requests. Looks good!

**Review by CarsonPruitt-NOAA — APPROVED** [review]

**Review by CarsonPruitt-NOAA — APPROVED** [review]

**Comment by mluck** [comment]
Does this close #212?

**Pull request merged**